src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp

```diff
@@ -378,9 +378,182 @@ void PDPServer2::announceParticipantState(
         bool dispose /* = false */,
         WriteParams& )
 {
-    (void)new_change;
-    (void)dispose;
-    // TODO DISCOVERY SERVER VERSION 2
+    logInfo(RTPS_PDP, "Announcing Server (new change: " << new_change << ")");
+    CacheChange_t* change = nullptr;
+
+    StatefulWriter* pW = dynamic_cast<StatefulWriter*>(mp_PDPWriter);
+    assert(pW);
+
+    /*
+       Protect writer sequence number. Make sure in order to prevent AB BA deadlock that the
+       writer mutex is systematically lock before the PDP one (if needed):
+        - transport callbacks on PDPListener
+        - initialization and removal on BuiltinProtocols::initBuiltinProtocols and ~BuiltinProtocols
+        - DSClientEvent (own thread)
+        - ResendParticipantProxyDataPeriod (participant event thread)
+     */
+    std::lock_guard<fastrtps::RecursiveTimedMutex> wlock(pW->getMutex());
+
+    if (!dispose)
+    {
+        // Create the CacheChange_t if necessary
+        if (m_hasChangedLocalPDP.exchange(false) || new_change)
+        {
+            getMutex()->lock();
+
+            // Copy the participant data
+            ParticipantProxyData proxy_data_copy(*getLocalParticipantProxyData());
+
+            // Prepare identity
+            WriteParams wp;
+            // TODO: Make sure in the database that this CacheChange_t is given this sequence number
```

Comment:
The own Data(P) is storaged with key GUIDPrefix, so this change will be safely storaged and recognized by the ddb.